src/java/org/apache/cassandra/cql3/Lists.java

```diff
@@ -492,26 +492,40 @@ public void execute(DecoratedKey partitionKey, UpdateParameters params) throws I
 
         static void doAppend(Term.Terminal value, ColumnMetadata column, UpdateParameters params) throws InvalidRequestException
         {
-            if (column.type.isMultiCell())
+            if (value == null)
             {
+                // for frozen lists, we're overwriting the whole cell value
+                if (!column.type.isMultiCell())
+                    params.addTombstone(column);
+
                 // If we append null, do nothing. Note that for Setter, we've
                 // already removed the previous value so we're good here too
-                if (value == null)
-                    return;
+                return;
+            }
 
-                for (ByteBuffer buffer : ((Value) value).elements)
+            List<ByteBuffer> elements = ((Value) value).elements;
+
+            if (column.type.isMultiCell())
+            {
+                // Guardrails about collection size are only checked for the added elements without considering
+                // already existent elements. This is done so to avoid read-before-write, having additional checks
+                // during SSTable write.
+                Guardrails.itemsPerCollection.guard(elements.size(), column.name.toString(), params.clientState);
```

Comment:
Isn't this in the hot path? You could return directly on 0 elements and spare 2 guardrails calls and a loop despite being 0 itearions

Reply:
Good idea, +1